implementations/utils/gpu.py

```diff
@@ -0,0 +1,386 @@
+    ssh_keys: airflow.XComArg,
+) -> str:
+  """Request a resource and wait until the nodes are created.
+
+  Args:
+    gpu_name: XCom value for unique GPU name.
+    image_project: project of the image.
+    image_family: family of the image.
+    accelerator: Description of GPU to create.
+    gcp: GCP project/zone configuration.
+    ssh_keys: XCom value for SSH keys to communicate with these GPUs.
+
+  Returns:
+    The ip address of the GPU VM.
+  """
+  image = get_image_from_family(project=image_project, family=image_family)
+  disk_type = f"zones/{gcp.zone}/diskTypes/pd-standard"
+  disks = [disk_from_image(disk_type, 100, True, image.self_link)]
+  metadata = create_metadata({
+      # "install-nvidia-driver": "True",
+      "install-nvidia-driver": "False",
+      "proxy-mode": "project_editors",
+      "ssh-keys": f"cloud-ml-auto-solutions:{ssh_keys.public}",
+  })
+  acceleratorConfig = compute_v1.AcceleratorConfig(
+      accelerator_count=accelerator.count,
+      accelerator_type=f"projects/{gcp.project_name}/zones/{gcp.zone}/acceleratorTypes/{accelerator.accelerator_type}",
+  )
+  service_account = compute_v1.ServiceAccount(
+      # email = "cloud-auto-ml-solutions@google.com",
+      scopes=["https://www.googleapis.com/auth/cloud-platform"]
+  )
+  instance = create_instance(
+      project_id=gcp.project_name,
+      zone=gcp.zone,
+      instance_name=gpu_name,
+      disks=disks,
+      machine_type=accelerator.machine_type,
+      service_account=service_account,
+      external_access=True,
+      accelerators=[acceleratorConfig],
+      metadata=metadata,
+      instance_termination_action="DELETE",
+  )
+  logging.info("instance info: {instance}")
+
+  ip_pattern = re.compile(r'network_i_p:\s+"([^"]+)"')
+  match = ip_pattern.search(str(instance))
+
+  # Extract the matched IP address
+  ip_address = "0.0.0.0"
+  if match:
+    ip_address = match.group(1)
+    logging.info(f"Created vm with ip address {ip_address}.")
+  else:
+    logging.error(f"No IP address found for instance: {gpu_name}.")
+    raise (f"Failed to create GPU resource {gpu_name}.")
+  return ip_address
+
+
+@task
+def ssh_host(ip_address: str, cmds: Iterable[str], ssh_keys: ssh.SshKeys) -> None:
+  """SSH GPU and run commands in multi process.
+
+  Args:
+   ip_address: The ip address of the vm resource.
+   cmds: The commands to run on a GPU.
+   ssh_keys: The SSH key pair to use for authentication.
+  """
+  pkey = paramiko.RSAKey.from_private_key(io.StringIO(ssh_keys.private))
+  logging.info(f"Connecting to IP addresses {ip_address}")
+
+  ssh_group = fabric.ThreadingGroup(
+      ip_address,
+      user="cloud-ml-auto-solutions",
+      connect_kwargs={
+          "auth_strategy": paramiko.auth_strategy.InMemoryPrivateKey(
+              "cloud-ml-auto-solutions", pkey
+          )
+      },
+  )
+  ssh_group.run(cmds)
+
+
+@task(trigger_rule="all_done")
+def delete_resource(instance_name: str, project_id: str, zone: str) -> str:
```

Comment:
I agree with Will. It's better we have a func called delete_resource_request, and sensor called wait_for_resource_deletion to be consistent with existing implementation. https://screenshot.googleplex.com/8uMfXiEXzoFqv62

Reply:
Based on the current workflow, what will be the use cases for:

1) we set `instance_termination_action` as stop, and then clean it up at the end
2) we set `instance_termination_action` as delete, and then clean it up at the end

Do we need this clean up? or do we need to add `instance_termination_action` param. Thanks?

Reply:
Updated to use STOP as the termination action since there is resource_deletion in the cleanup stage.